llrt_core/src/modules/encoding/text_decoder.rs

```diff
@@ -65,25 +60,27 @@ impl<'js> TextDecoder {
 
     pub fn decode(&self, ctx: Ctx<'js>, buffer: Value<'js>) -> Result<String> {
         let bytes = get_bytes(&ctx, buffer)?;
-
-        let decoder = Encoding::for_label(self.encoding.as_bytes()).unwrap();
-
-        let str: Cow<str>;
-        let has_error: bool;
-
-        if decoder == encoding_rs::UTF_8 {
-            (str, has_error) = match self.ignore_bom {
-                false => decoder.decode_with_bom_removal(&bytes),
-                true => decoder.decode_without_bom_handling(&bytes),
-            }
+        let start_pos = if !self.ignore_bom && bytes.len() >= 2 && bytes[..2] == [0xFF, 0xFE] {
+            2
+        } else if !self.ignore_bom && bytes.len() >= 3 && bytes[..3] == [0xEF, 0xBB, 0xBF] {
+            3
         } else {
-            (str, _, has_error) = decoder.decode(&bytes);
-        }
-
-        if self.fatal && has_error {
-            return Err(Exception::throw_message(&ctx, "Fatal error"));
+            0
+        };
+
+        let str = self
+            .encoder
+            .encode_to_string(&bytes[start_pos..])
+            .or_throw(&ctx)?;
+
+        if self.fatal
+            && str
+                .as_bytes()
+                .windows(3)
+                .any(|window: &[u8]| window == [0xEF, 0xBF, 0xBD])
+        {
+            return Err(Exception::throw_type(&ctx, "Fatal error"));
         }
```

Comment:
We're already checking the string and replacing unknown chars to � in the https://github.com/nabetti1720/llrt/blob/8618194fafd85fc16eed82e2be0a6588d2338897/llrt_core/src/modules/encoding/encoder.rs#L113 method. It would be more optimal to not do `String::from_utf8_lossy` if `self.fatal`. This avoids iterating through the string twice.

Reply:
Indeed it is! I have corrected it and would appreciate a review.